### ISSUE TITLE
Multi-GPUs with optuna

### ISSUE THREAD

**Issue by Uproar197** [optuna/optuna#2286]
Hello, now I meet a problem, the overview of my code is shown in the following link:
https://github.com/Uproar197/Code/blob/main/my_optuna_problem.py

I can use Optuna to conduct experiments, but I don’t know why when the training epoch is relatively small, eg. **5 epochs**,  the experiment can be performed according to the given trial value and output; but when the number of the epoch is relatively large, eg. **250 epochs**,  no matter how big **n_trial** is,  the experiment comes the fifth trial, it will **be stuck** and cannot continue.

Another question is that when I use 5 GPUs, it will start 5 trials during the experiment, but it seems that these 5 trials are only equivalent to one trial in n_trials. I'm not sure why?
![image](https://user-images.githubusercontent.com/46192007/106249841-33298280-6256-11eb-89ba-dd4691d4aa82.png)

In addition, I output each training process, but there is a problem after using optuna, that is, the number of outputs will increase as the number of trials increases. For example, in the **first trial**, **'Epoch: [{}/{}] Iter:[{}/{}], Time: {:.2f}, lr: {:.6f}, Loss: {:.6f}'** will output once; but when it comes to the **third trial**,  **'Epoch: [{}/{}] Iter:[{}/{}], Time: {:.2f}, lr: {:.6f}, Loss: {:.6f}'** will output three times. Do you know the reason?
![image](https://user-images.githubusercontent.com/46192007/106250444-fdd16480-6256-11eb-908c-43c74a4d7942.png)
![image](https://user-images.githubusercontent.com/46192007/106250534-1c376000-6257-11eb-9abb-3ffc5472ff2c.png)

Thanks in advance!

**Comment by nzw0301** [comment]
I would answer one of your questions for sure.

> Another question is that when I use 5 GPUs, it will start 5 trials during the experiment, but it seems that these 5 trials are only equivalent to one trial in n_trials. I'm not sure why?

This is because you use PyTorch's distributed data-parallel.

I think other questions are also related to distributed data-parallel training.

**Comment by Uproar197** [comment]
Sorry for my late response. Maybe it is the case you answered. And is there anyone who knows why it will be stuck?

**Comment by nzw0301** [comment]
I think that the deadlock of Sqlite is the reason that your code stops. The other databases, such as MySQL or PostgreSQL, can prevent your issue. Please see details in https://optuna.readthedocs.io/en/latest/tutorial/10_key_features/004_distributed.html

**Comment by Uproar197** [comment]
Thanks a lot! I will check the link you offer to see if it is the reason.

**Comment by Uproar197** [comment]
Unfortunately，it seems not about the database. I changed the database from Sqlite to MySQL and ran the codes, it's still stuck when the trial number comes to five.

**Comment by nzw0301** [comment]
Hi, @Uproar197 sorry for confusing you by answering the wrong approach. I think you can solve your issue by using a new feature introduced by #2303.

**Comment by Crissman** [comment]
Since there hasn't been recent activity on this, I assume this issue is resolved and will close this issue. Please feel free to re-open if still an issue.
